src/main.cpp

```diff
@@ -1145,6 +1186,7 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
                 pfrom->fDisconnect = true;
             }
         }
+        pfrom->state_incoming = ConnectionStateIncoming::SENT_VERACK_READY_FOR_POTENTIAL_XVERSION;
```

Comment:
This must be set before "pfrom->PushMessage(NetMsgType::VERACK);" or we could legitimately receive the XVERSION response before we set this state flag due to the multithreaded network processing.